timelock-impl/src/main/java/com/palantir/atlasdb/timelock/management/AllNodesDisabledNamespacesUpdater.java

```diff
@@ -0,0 +1,320 @@
+        // to avoid leaving ourselves in an inconsistent state.
+        boolean rollbackSuccess = attemptReEnableOnAllNodes(namespaces, lockId, responses.size());
+        if (rollbackSuccess) {
+            log.error(
+                    "Failed to disable all namespaces. However, we successfully rolled back any partially disabled"
+                            + " namespaces.",
+                    SafeArg.of("namespaces", namespaces),
+                    SafeArg.of("lockId", lockId));
+            return DisableNamespacesResponse.unsuccessful(
+                    UnsuccessfulDisableNamespacesResponse.builder().build());
+        }
+
+        log.error(
+                "Failed to disable all namespaces, and we failed to roll back some namespaces."
+                        + " These will need to be force-re-enabled in order to return Timelock to a consistent state.",
+                SafeArg.of("namespaces", namespaces),
+                SafeArg.of("lockId", lockId));
+        return DisableNamespacesResponse.unsuccessful(UnsuccessfulDisableNamespacesResponse.builder()
+                .partiallyDisabledNamespaces(namespaces)
+                .build());
+    }
+
+    public ReenableNamespacesResponse reEnableOnAllNodes(ReenableNamespacesRequest request) {
+        Set<Namespace> namespaces = request.getNamespaces();
+
+        if (anyNodeIsUnreachable()) {
+            log.error(
+                    "Failed to reach all remote nodes. Not re-enabling any namespaces",
+                    SafeArg.of("namespaces", namespaces));
+            return ReenableNamespacesResponse.unsuccessful(
+                    UnsuccessfulReenableNamespacesResponse.builder().build());
+        }
+
+        List<SingleNodeUpdateResponse> responses = reEnableNamespacesOnAllNodes(request);
+        if (updateWasSuccessfulOnAllNodes(responses)) {
+            return ReenableNamespacesResponse.successful(SuccessfulReenableNamespacesResponse.of(true));
+        }
+
+        Set<Namespace> consistentlyLockedNamespaces = getConsistentFailures(responses);
+        if (!consistentlyLockedNamespaces.isEmpty()) {
+            log.error(
+                    "Failed to re-enable all namespaces, because some namespace was consistently disabled "
+                            + "with the wrong lock ID. This implies that this namespace being restored by another"
+                            + " process. If that is the case, please either wait for that restore to complete, "
+                            + " or kick off a restore without that namespace",
+                    SafeArg.of("lockedNamespaces", consistentlyLockedNamespaces));
+            return ReenableNamespacesResponse.unsuccessful(UnsuccessfulReenableNamespacesResponse.builder()
+                    .consistentlyLockedNamespaces(consistentlyLockedNamespaces)
+                    .build());
+        }
+
+        UUID lockId = request.getLockId();
+        // should only roll back locally
```

Comment:
* should not roll back locally - if we're here, there's no situation where we successfully made the local update.

For the remote ones, we should try to roll back all of them (e.g. if we failed, we might have failed because of a timeout or something).